File: example/java/jimmer-sql-graphql/runtime/src/main/java/org/babyfish/jimmer/sql/example/runtime/filter/TenantFilterForNonCacheMode.java
package org.babyfish.jimmer.sql.example.runtime.filter;

import org.babyfish.jimmer.sql.example.model.common.TenantAwareProps;
import org.babyfish.jimmer.sql.example.runtime.TenantProvider;
import org.babyfish.jimmer.sql.filter.Filter;
import org.babyfish.jimmer.sql.filter.FilterArgs;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.stereotype.Component;

/*
 * see JSqlClient.Builder.addFilters
 * 
 * This bean is only be used when cache is NOT used.
 */
@ConditionalOnMissingBean(TenantFilterForCacheMode.class)
@Component
public class TenantFilterForNonCacheMode implements Filter<TenantAwareProps> { // ❶

    protected final TenantProvider tenantProvider;

    public TenantFilterForNonCacheMode(TenantProvider tenantProvider) {
        this.tenantProvider = tenantProvider;
    }

    @Override
    public void filter(FilterArgs<TenantAwareProps> args) {
        String tenant = tenantProvider.get();
        if (tenant != null) {
            args.where(args.getTable().tenant().eq(tenant));
        }
    }
}

/*----------------Documentation Links----------------
❶ https://babyfish-ct.github.io/jimmer/docs/query/global-filter/user-filter
---------------------------------------------------*/

File: example/kotlin/jimmer-sql-kt/runtime/src/main/kotlin/org/babyfish/jimmer/sql/example/kt/runtime/filter/TenantFilterForNonCacheMode.kt
package org.babyfish.jimmer.sql.example.kt.runtime.filter

import org.babyfish.jimmer.sql.example.kt.runtime.TenantProvider
import org.babyfish.jimmer.sql.example.kt.model.common.TenantAware
import org.babyfish.jimmer.sql.example.kt.model.common.tenant
import org.babyfish.jimmer.sql.kt.ast.expression.eq
import org.babyfish.jimmer.sql.kt.filter.KFilter
import org.babyfish.jimmer.sql.kt.filter.KFilterArgs
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.stereotype.Component

/*
 * see KSqlClientDsl.addFilters
 *
 * This bean is only be used when cache is NOT enabled.
 */
@ConditionalOnMissingBean(TenantFilterForCacheMode::class)
@Component
open class TenantFilterForNonCacheMode(
    protected val tenantProvider: TenantProvider
) : KFilter<TenantAware> { // ❶

    override fun filter(args: KFilterArgs<TenantAware>) {
        tenantProvider.tenant?.let {
            args.apply {
                where(table.tenant eq it)
            }
        }
    }
}

/*----------------Documentation Links----------------
❶ https://babyfish-ct.github.io/jimmer/docs/query/global-filter/user-filter
---------------------------------------------------*/
